File: admin/scripts/import-tech-terms.py
#!/usr/bin/env python3
"""
Import Technical Terms from CSV to Supabase

This script imports your existing 593 technical terms from CSV into the
tech_terms table in Supabase.

Usage:
    python3 import-tech-terms.py /path/to/csv/file.csv
"""

import csv
import sys
import os
from supabase import create_client, Client

# ============================================================================
# Configuration
# ============================================================================

CSV_FILE = sys.argv[1] if len(sys.argv) > 1 else '/Users/keola/Downloads/mk_tech_terms_trilingual_haw_extracted.csv'

# Load from environment variables or admin/supabase/.env
SUPABASE_URL = os.getenv('SUPABASE_URL', '')
SUPABASE_KEY = os.getenv('SUPABASE_ANON_KEY', '')

# Try to load from .env file if not in environment
if not SUPABASE_URL or not SUPABASE_KEY:
    env_file = '/Users/keola/Documents/GitHub/RMHokeoUnuhina/admin/supabase/.env'
    if os.path.exists(env_file):
        with open(env_file, 'r') as f:
            for line in f:
                if line.startswith('SUPABASE_URL='):
                    SUPABASE_URL = line.split('=', 1)[1].strip()
                elif line.startswith('SUPABASE_ANON_KEY='):
                    SUPABASE_KEY = line.split('=', 1)[1].strip()

# ============================================================================
# CSV Reading
# ============================================================================

def read_csv(file_path):
    """Read CSV file and return list of dictionaries"""
    rows = []
    with open(file_path, 'r', encoding='utf-8-sig') as f:
        reader = csv.DictReader(f)
        for row in reader:
            rows.append(row)
    return rows

# ============================================================================
# Data Transformation
# ============================================================================

def map_review_status(csv_status):
    """Map CSV review_status to new status values"""
    status_map = {
        'ok': 'approved',
        'needs_check': 'pending_approval',
        'unpaired': 'pending_approval'
    }
    return status_map.get(csv_status, 'pending_approval')

def map_overall_review_status(csv_status):
    """Map to overall review status"""
    status_map = {
        'ok': 'ok',
        'needs_check': 'needs_check',
        'unpaired': 'needs_check'
    }
    return status_map.get(csv_status, 'needs_check')

def transform_row(csv_row):
    """Transform CSV row to database row format"""
    return {
        # English
        'eng_term': csv_row.get('eng_term') or None,
        'eng_definition': csv_row.get('eng_definition') or None,
        'eng_pos': csv_row.get('eng_pos') or None,
        'eng_notes': csv_row.get('eng_notes') or None,

        # Hawaiian
        'haw_term': csv_row.get('haw_term') or None,
        'haw_definition': csv_row.get('haw_definition') or None,
        'haw_pos': csv_row.get('haw_pos') or None,
        'haw_notes': csv_row.get('haw_notes') or None,
        'haw_source': csv_row.get('source') or None,
        'haw_status': map_review_status(csv_row.get('review_status', '')),

        # Māori
        'mao_term': csv_row.get('mao_term') or None,
        'mao_definition': csv_row.get('mao_definition') or None,
        'mao_pos': csv_row.get('mao_pos') or None,
        'mao_notes': csv_row.get('mao_notes') or None,
        'mao_source': csv_row.get('source') or None,
        'mao_status': map_review_status(csv_row.get('review_status', '')),

        # Metadata
        'pair_id': csv_row.get('pair_id') or None,
        'review_status': map_overall_review_status(csv_row.get('review_status', '')),
        'primary_source': csv_row.get('source') or None,

        # Domain - to be filled later
        'primary_domain': None,
        'subdomain': None
    }

# ============================================================================
# Supabase Import
# ============================================================================

def import_to_supabase(rows):
    """Import rows to Supabase in batches"""
    supabase: Client = create_client(SUPABASE_URL, SUPABASE_KEY)

    print(f"\nImporting {len(rows)} terms to Supabase...")
    print("=========================================\n")

    success_count = 0
    error_count = 0
    errors = []

    # Import in batches of 50
    BATCH_SIZE = 50

    for i in range(0, len(rows), BATCH_SIZE):
        batch = rows[i:i + BATCH_SIZE]
        transformed_batch = [transform_row(row) for row in batch]

        try:
            result = supabase.table('tech_terms').insert(transformed_batch).execute()
            success_count += len(batch)
            batch_num = (i // BATCH_SIZE) + 1
            print(f"✓ Batch {batch_num}: Imported {len(batch)} terms ({success_count}/{len(rows)})")
        except Exception as e:
            error_count += len(batch)
            batch_num = (i // BATCH_SIZE) + 1
            error_msg = str(e)
            print(f"✗ Batch {batch_num} error: {error_msg}")
            errors.append({'batch': batch_num, 'error': error_msg})

    print('\n=========================================')
    print('Import Summary:')
    print('=========================================')
    print(f'Total rows processed: {len(rows)}')
    print(f'Successfully imported: {success_count}')
    print(f'Errors: {error_count}')

    if errors:
        print('\nError Details:')
        for err in errors:
            print(f"  Batch {err['batch']}: {err['error']}")

    return {'success_count': success_count, 'error_count': error_count, 'errors': errors}

# ============================================================================
# Statistics
# ============================================================================

def show_statistics():
    """Show database statistics after import"""
    supabase: Client = create_client(SUPABASE_URL, SUPABASE_KEY)

    print('\n=========================================')
    print('Database Statistics:')
    print('=========================================\n')

    # Total count
    result = supabase.table('tech_terms').select('*', count='exact').execute()
    total_count = result.count

    print(f'Total terms in database: {total_count}')

    # Hawaiian translation count
    haw_result = supabase.table('tech_terms').select('*', count='exact').not_.is_('haw_term', 'null').neq('haw_term', '').execute()
    haw_count = haw_result.count

    print(f'Terms with Hawaiian: {haw_count} ({(haw_count/total_count*100):.1f}%)')

    # Māori translation count
    mao_result = supabase.table('tech_terms').select('*', count='exact').not_.is_('mao_term', 'null').neq('mao_term', '').execute()
    mao_count = mao_result.count

    print(f'Terms with Māori: {mao_count} ({(mao_count/total_count*100):.1f}%)')

    # Status breakdown
    status_result = supabase.table('tech_terms').select('haw_status').execute()
    status_counts = {}
    for row in status_result.data:
        status = row['haw_status']
        status_counts[status] = status_counts.get(status, 0) + 1

    print('\nHawaiian Status Breakdown:')
    for status, count in status_counts.items():
        print(f'  {status}: {count}')

# ============================================================================
# Main
# ============================================================================

def main():
    print('=========================================')
    print('Technical Terms CSV Import')
    print('=========================================\n')

    # Check configuration
    if not SUPABASE_URL or not SUPABASE_KEY:
        print('ERROR: Please configure SUPABASE_URL and SUPABASE_KEY')
        print('You can set them as environment variables or in admin/supabase/.env')
        print('\nExample:')
        print('  export SUPABASE_URL="https://your-project.supabase.co"')
        print('  export SUPABASE_ANON_KEY="your-anon-key"')
        print('  python3 import-tech-terms.py')
        sys.exit(1)

    # Check CSV file
    if not os.path.exists(CSV_FILE):
        print(f'ERROR: CSV file not found: {CSV_FILE}')
        print('\nUsage:')
        print('  python3 import-tech-terms.py /path/to/file.csv')
        sys.exit(1)

    print(f'Reading CSV: {CSV_FILE}\n')

    try:
        # Parse CSV
        rows = read_csv(CSV_FILE)
        print(f'Parsed {len(rows)} rows from CSV\n')

        # Show sample
        print('Sample row (first entry):')
        import json
        print(json.dumps(rows[0], indent=2, ensure_ascii=False))
        print('')

        # Confirm before import
        answer = input('Proceed with import? (yes/no): ')

        if answer.lower() != 'yes':
            print('Import cancelled.')
            sys.exit(0)

        # Import
        result = import_to_supabase(rows)

        # Show statistics
        if result['success_count'] > 0:
            show_statistics()

        print('\nImport complete!')

    except Exception as e:
        print(f'ERROR: {e}')
        import traceback
        traceback.print_exc()
        sys.exit(1)

if __name__ == '__main__':
    main()

File: admin/scripts/import-tech-terms.js
#!/usr/bin/env node
/**
 * Import Technical Terms from CSV to Supabase
 *
 * This script imports your existing 593 technical terms from CSV into the
 * tech_terms table in Supabase.
 *
 * Usage:
 *   node import-tech-terms.js /path/to/csv/file.csv
 */

const fs = require('fs');
const { createClient } = require('@supabase/supabase-js');

// ============================================================================
// Configuration
// ============================================================================

const CSV_FILE = process.argv[2] || '/Users/keola/Downloads/mk_tech_terms_trilingual_haw_extracted.csv';

// Supabase configuration (loaded from environment or config)
const SUPABASE_URL = process.env.SUPABASE_URL || 'YOUR_SUPABASE_URL';
const SUPABASE_KEY = process.env.SUPABASE_ANON_KEY || 'YOUR_SUPABASE_KEY';

// ============================================================================
// CSV Parsing
// ============================================================================

function parseCSV(filePath) {
  const content = fs.readFileSync(filePath, 'utf-8');
  const lines = content.split('\n');

  // Remove BOM if present
  const firstLine = lines[0].replace(/^\uFEFF/, '');
  const headers = firstLine.split(',');

  const rows = [];
  for (let i = 1; i < lines.length; i++) {
    if (!lines[i].trim()) continue; // Skip empty lines

    const values = parseCSVLine(lines[i]);
    if (values.length !== headers.length) {
      console.warn(`Row ${i} has ${values.length} values but expected ${headers.length}`);
      continue;
    }

    const row = {};
    headers.forEach((header, index) => {
      row[header.trim()] = values[index];
    });
    rows.push(row);
  }

  return rows;
}

function parseCSVLine(line) {
  const values = [];
  let current = '';
  let inQuotes = false;

  for (let i = 0; i < line.length; i++) {
    const char = line[i];

    if (char === '"') {
      inQuotes = !inQuotes;
    } else if (char === ',' && !inQuotes) {
      values.push(current.trim());
      current = '';
    } else {
      current += char;
    }
  }
  values.push(current.trim());

  return values;
}

// ============================================================================
// Data Transformation
// ============================================================================

function transformRow(csvRow) {
  // Map CSV column names to database column names
  return {
    // English
    eng_term: csvRow.eng_term || null,
    eng_definition: csvRow.eng_definition || null,
    eng_pos: csvRow.eng_pos || null,
    eng_notes: csvRow.eng_notes || null,
    // eng_examples: Currently empty in CSV

    // Hawaiian
    haw_term: csvRow.haw_term || null,
    haw_definition: csvRow.haw_definition || null,
    haw_pos: csvRow.haw_pos || null,
    haw_notes: csvRow.haw_notes || null,
    haw_source: csvRow.source || null, // Primary source goes to Hawaiian
    haw_status: mapReviewStatus(csvRow.review_status),
    // haw_examples: Currently empty in CSV

    // Māori
    mao_term: csvRow.mao_term || null,
    mao_definition: csvRow.mao_definition || null,
    mao_pos: csvRow.mao_pos || null,
    mao_notes: csvRow.mao_notes || null,
    mao_source: csvRow.source || null, // Same source for now
    mao_status: mapReviewStatus(csvRow.review_status),
    // mao_examples: Currently empty in CSV

    // Metadata
    pair_id: csvRow.pair_id || null,
    review_status: mapOverallReviewStatus(csvRow.review_status),
    primary_source: csvRow.source || null,

    // Domain - will need to be added later manually or via separate script
    primary_domain: null,
    subdomain: null
  };
}

function mapReviewStatus(csvStatus) {
  // Map CSV review_status to our new status values
  const statusMap = {
    'ok': 'approved',
    'needs_check': 'pending_approval',
    'unpaired': 'pending_approval'
  };

  return statusMap[csvStatus] || 'pending_approval';
}

function mapOverallReviewStatus(csvStatus) {
  // Map to overall review status
  const statusMap = {
    'ok': 'ok',
    'needs_check': 'needs_check',
    'unpaired': 'needs_check'
  };

  return statusMap[csvStatus] || 'needs_check';
}

// ============================================================================
// Supabase Import
// ============================================================================

async function importToSupabase(rows) {
  const supabase = createClient(SUPABASE_URL, SUPABASE_KEY);

  console.log(`\nImporting ${rows.length} terms to Supabase...`);
  console.log('=========================================\n');

  let successCount = 0;
  let errorCount = 0;
  const errors = [];

  // Import in batches of 50
  const BATCH_SIZE = 50;

  for (let i = 0; i < rows.length; i += BATCH_SIZE) {
    const batch = rows.slice(i, i + BATCH_SIZE);
    const transformedBatch = batch.map(transformRow);

    const { data, error } = await supabase
      .from('tech_terms')
      .insert(transformedBatch);

    if (error) {
      console.error(`Batch ${Math.floor(i/BATCH_SIZE) + 1} error:`, error.message);
      errorCount += batch.length;
      errors.push({ batch: Math.floor(i/BATCH_SIZE) + 1, error: error.message });
    } else {
      successCount += batch.length;
      console.log(`✓ Batch ${Math.floor(i/BATCH_SIZE) + 1}: Imported ${batch.length} terms (${successCount}/${rows.length})`);
    }
  }

  console.log('\n=========================================');
  console.log('Import Summary:');
  console.log('=========================================');
  console.log(`Total rows processed: ${rows.length}`);
  console.log(`Successfully imported: ${successCount}`);
  console.log(`Errors: ${errorCount}`);

  if (errors.length > 0) {
    console.log('\nError Details:');
    errors.forEach(err => {
      console.log(`  Batch ${err.batch}: ${err.error}`);
    });
  }

  return { successCount, errorCount, errors };
}

// ============================================================================
// Statistics
// ============================================================================

async function showStatistics() {
  const supabase = createClient(SUPABASE_URL, SUPABASE_KEY);

  console.log('\n=========================================');
  console.log('Database Statistics:');
  console.log('=========================================\n');

  // Total count
  const { count: totalCount } = await supabase
    .from('tech_terms')
    .select('*', { count: 'exact', head: true });

  console.log(`Total terms in database: ${totalCount}`);

  // Hawaiian translation count
  const { count: hawCount } = await supabase
    .from('tech_terms')
    .select('*', { count: 'exact', head: true })
    .not('haw_term', 'is', null)
    .neq('haw_term', '');

  console.log(`Terms with Hawaiian: ${hawCount} (${((hawCount/totalCount)*100).toFixed(1)}%)`);

  // Māori translation count
  const { count: maoCount } = await supabase
    .from('tech_terms')
    .select('*', { count: 'exact', head: true })
    .not('mao_term', 'is', null)
    .neq('mao_term', '');

  console.log(`Terms with Māori: ${maoCount} (${((maoCount/totalCount)*100).toFixed(1)}%)`);

  // Status breakdown
  const { data: statusData } = await supabase
    .from('tech_terms')
    .select('haw_status');

  const statusCounts = statusData.reduce((acc, row) => {
    acc[row.haw_status] = (acc[row.haw_status] || 0) + 1;
    return acc;
  }, {});

  console.log('\nHawaiian Status Breakdown:');
  Object.entries(statusCounts).forEach(([status, count]) => {
    console.log(`  ${status}: ${count}`);
  });
}

// ============================================================================
// Main
// ============================================================================

async function main() {
  console.log('=========================================');
  console.log('Technical Terms CSV Import');
  console.log('=========================================\n');

  // Check configuration
  if (SUPABASE_URL === 'YOUR_SUPABASE_URL' || SUPABASE_KEY === 'YOUR_SUPABASE_KEY') {
    console.error('ERROR: Please configure SUPABASE_URL and SUPABASE_KEY');
    console.error('You can set them as environment variables or edit this script.');
    console.error('\nExample:');
    console.error('  export SUPABASE_URL="https://your-project.supabase.co"');
    console.error('  export SUPABASE_ANON_KEY="your-anon-key"');
    console.error('  node import-tech-terms.js');
    process.exit(1);
  }

  // Check CSV file
  if (!fs.existsSync(CSV_FILE)) {
    console.error(`ERROR: CSV file not found: ${CSV_FILE}`);
    console.error('\nUsage:');
    console.error('  node import-tech-terms.js /path/to/file.csv');
    process.exit(1);
  }

  console.log(`Reading CSV: ${CSV_FILE}\n`);

  try {
    // Parse CSV
    const rows = parseCSV(CSV_FILE);
    console.log(`Parsed ${rows.length} rows from CSV\n`);

    // Show sample
    console.log('Sample row (first entry):');
    console.log(JSON.stringify(rows[0], null, 2));
    console.log('');

    // Confirm before import
    const readline = require('readline').createInterface({
      input: process.stdin,
      output: process.stdout
    });

    readline.question('Proceed with import? (yes/no): ', async (answer) => {
      readline.close();

      if (answer.toLowerCase() !== 'yes') {
        console.log('Import cancelled.');
        process.exit(0);
      }

      // Import
      const result = await importToSupabase(rows);

      // Show statistics
      if (result.successCount > 0) {
        await showStatistics();
      }

      console.log('\nImport complete!');
    });

  } catch (error) {
    console.error('ERROR:', error.message);
    console.error(error.stack);
    process.exit(1);
  }
}

// Run if called directly
if (require.main === module) {
  main();
}

module.exports = { parseCSV, transformRow, importToSupabase };
